### ISSUE TITLE
[JENKINS-4008] Memory leak when building on slaves

### ISSUE THREAD

**Issue by timja** [timja/jenkins-gh-issues-poc-06-20#63]
<p>Hello,<br/>
I strongly suspect there's a memory leak in Hudson concerning building on<br/>
slaves. It may be related to issue 3076, and certainly to<br/>
<a href="http://www.nabble.com/Possible-memory-leak-in-hudson.remoting.ExportTable-td12000299.html" class="external-link" target="_blank" rel="nofollow noopener">http://www.nabble.com/Possible-memory-leak-in-hudson.remoting.ExportTable-td12000299.html</a>,<br/>
but it's not entirely the same.<br/>
We run Hudson 314 on Windows XP, with master and slaves on the same machine. I<br/>
see the same symptoms whether the slaves are started by the master or started<br/>
via jnlp.<br/>
The best numbers I have are from a week of running 313 (slaves started by<br/>
master), where I had java log GC activity to a file. Grepping "Full GC" lines<br/>
out of the log and massaging the numbers into a spreadsheet, a clear linear<br/>
growth in memory consumption is seen. From those numbers I could predict that we<br/>
would run out of heap (512M) sometime on day 24 of running.<br/>
This morning I switched to 314 (slaves start via jnlp). The first day gave<br/>
essentially flat memory curve, but it's starting to grow by the end of the day.<br/>
It's clear from dumpExportTable that each and every build leaks objects. They do<br/>
not disappear when the build ends, nor does a </p>
{hudson_url}
<p>/gc remove them. This<br/>
was observed on a slave ("stager", see below) with builds that where started<br/>
manually by me. No SCM or triggering involved (the builds would normally be<br/>
triggered by upstream builds).<br/>
But the week of logging GC also demonstrates that the growth in memory is the<br/>
same over the weekend and during the nights, when essentially no builds took<br/>
place. So something else is at play here; perhaps the SCM polling? We use the<br/>
Perforce plugin for that.<br/>
I have dumpExportTable dumps from all three slaves (see attached). The first<br/>
("builder") has 3 projects, all polling SCM. Builds take a long time, so it can<br/>
perform perhaps 10 builds in one day. The dump is 5428 KB. The second ("tester")<br/>
runs the unittests of 2 of the builder projects. They take even longer, so it<br/>
runs perhaps 6-7 builds in a day. Triggering is done by upstream, but SCM is<br/>
involved to sync the right files. That dump is 28K. The third ("stager") copies<br/>
artifacts from a number of builds (both from "builder" and "master") to a<br/>
staging area. That's somewhat frequent, and is only triggered by upstream<br/>
although part of the staging is done from the SCM. This dump is 36K.<br/>
Both "builder" and "tester" will regularly have 1-2 jobs in their queues at all<br/>
times.<br/>
The dumps are pretty much the same. The "builder" has these two exports<br/>
repeatedly (LOTS of them):<br/>
#3 (ref.1) : java.io.PipedInputStream@29830f<br/>
#4 (ref.1) : hudson.CloseProofOutputStream@12bde6c<br/>
The "tester" has no PipedInputStream, but only CloseProofOutputStream and a<br/>
RemoteClassLoader.<br/>
The "stager" only has CloseProofOutputStream.</p>

---
<details><summary><i>Originally reported by <a title="torbent" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=torbent">torbent</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-4008" target="_blank">Memory leak when building on slaves</a></i></summary>
<i><ul>
<li><b>status</b>: Resolved
<li><b>priority</b>: Critical
<li><b>resolution</b>: Cannot Reproduce
<li><b>resolved</b>: 2014-05-10T11:57:19+00:00
<li><b>imported</b>: 2022/01/10
</ul></i>
</details>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=torbent">torbent</a>:</i>
<p>Created an attachment (id=775)<br/>
Outputs of dumpExportTable from three slaves</p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=torbent">torbent</a>:</i>
<p>A possible leak probably warrants a P2?</p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=mindless">mindless</a>:</i>
<p>see my comment on related issue #3996</p>

<p>The CloseProofOutputStream items here seem to be the same problem as the<br/>
FileOutputStream items in 3996.  The PipedInputStream items seem to be similar,<br/>
but for an input stream.. here it looks like the unexport is supposed to happen<br/>
in ProxyInputStream.execute instead of finalize as done in ProxyOutputStream.</p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=mindless">mindless</a>:</i>
<ul>
	<li>
	<ul>
		<li>
		<ul>
			<li>Issue 3996 has been marked as a duplicate of this issue. ***</li>
		</ul>
		</li>
	</ul>
	</li>
</ul>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=mindless">mindless</a>:</i>
<p>Email from roberthanna:</p>

<p>I've also found in UserRequest class:</p>

<p>    public UserRequest(Channel local, Callable<?,EXC> c) throws<br/>
IOException {<br/>
exports = local.startExportRecording();<br/>
try </p>
{
    request = serialize(c,local);
    this.toString = c.toString();
}
<p> finally </p>
{
    exports.stopRecording();
}<br/>
ClassLoader cl = getClassLoader(c);<br/>
classLoaderProxy = RemoteClassLoader.export(cl,local);<br/>
    }<br/>
<br/>
Which should be:<br/>
<br/>
    public UserRequest(Channel local, Callable<?,EXC> c) throws<br/>
IOException {<br/>
exports = local.startExportRecording();<br/>
try {
    request = serialize(c,local);
    this.toString = c.toString();
    ClassLoader cl = getClassLoader(c);
    classLoaderProxy = RemoteClassLoader.export(cl,local);
} finally {    exports.stopRecording();}
<p>    }</p>

<p>So that the RemoteClassLoader gets unexported when the UserRequest gets<br/>
completed.<br/>
I've also had to refactor some of the code in the base Request class to<br/>
ensure that<br/>
the UserRequest objects are unexported when the job completes.</p>

<p>This has alleviated by a LOT the memory leak, but the master is still<br/>
leaking more exports.</p>

<p>When I have time, I'll look at the other exports remaining.</p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=mdillon">mdillon</a>:</i>
<p>I think this is a duplicate of 4045 (or rather vice versa).</p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=torbent">torbent</a>:</i>
<p>Thank you, mdillon. I'll recheck again when the fix for issue 4045 has been<br/>
released. It needs to run for a couple of days before I can verify.</p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=mdillon">mdillon</a>:</i>
<p>The fix in issue 4045 definitely fixes the issue with all the OutputStream<br/>
objects being retained, but I'm not sure about the RemoteClassLoader refcount<br/>
issue mentioned by mindeless. What I can say is that there is only one<br/>
RemoteClassLoader per remote node, so I doubt it would cause a memory leak even<br/>
if the reference count is not being tracked correctly (not saying it shouldn't<br/>
be fixed if it isn't correct).</p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=torbent">torbent</a>:</i>
<p>Have the changes that mindless and/or roberthanna refer to been integrated?<br/>
If so, that would take care of the RemoteClassLoaders ...<br/>
If not, perhaps they should?</p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kohsuke">kohsuke</a>:</i>
<p>Keeping ClassLoaders exported is a design choice. Otherwise we'll end up<br/>
creating a lots of ClassLoaders that load the classes of the same name,<br/>
resulting in mysterious ClassCastExceptions everywhere.</p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kohsuke">kohsuke</a>:</i>
<p>Closing this issue. # of ClassLoaders are not unbound, and given the feedback<br/>
after #4045 fix, I declare the issue resolved.</p>

<p>If people suspect other leaks, please feel free to reopen with the export table<br/>
dumps.</p>

<ul>
	<li>
	<ul>
		<li>
		<ul>
			<li>This issue has been marked as a duplicate of 4045 ***</li>
		</ul>
		</li>
	</ul>
	</li>
</ul>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=torbent">torbent</a>:</i>
<p>I've upgraded to 1.318 now, and the memory consumption rate is MUCH lower. I'm<br/>
sure it could run for months now before running out of memory.<br/>
But I still have many "leaked" objects of type CloseProofOutputStream in my<br/>
export tables, so I'm reopening this issue.<br/>
New dump will be attached.</p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=torbent">torbent</a>:</i>
<p>Created an attachment (id=835)<br/>
dumpExportTable dumps, same setup as earlier, Hudson freshly GCed</p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=oleg_nenashev">oleg_nenashev</a>:</i>
<p>I suppose the issue is very outdated due to many changes in Jenkins slaves.<br/>
remoting-2.32  works for months without significant memory leaks (at least at our use-cases) </p>

**Comment by timja** [comment]
<i>[Originally duplicated by: <a href="https://github.com/timja/jenkins-gh-issues-poc-06-20/issues?q=in%3Atitle%20JENKINS-3996">JENKINS-3996</a>]</i>
